programs/mmm/src/instructions/ext_vanilla/ext_fulfill_buy.rs

```diff
@@ -0,0 +1,335 @@
+
+        invoke_transfer_checked(
+            token_program.key,
+            payer_asset_account.to_account_info(),
+            asset_mint.to_account_info(),
+            owner_token_account.to_account_info(),
+            payer.to_account_info(),
+            remaining_accounts,
+            args.asset_amount,
+            0,   // decimals
+            &[], // seeds
+        )?;
+    }
+
+    // we can close the payer_asset_account if no amount left
+    if payer_asset_account.amount == args.asset_amount {
+        close_account(CpiContext::new(
+            token_program.to_account_info(),
+            anchor_spl::token_2022::CloseAccount {
+                account: payer_asset_account.to_account_info(),
+                destination: payer.to_account_info(),
+                authority: payer.to_account_info(),
+            },
+        ))?;
+    }
+
+    // prevent frontrun by pool config changes
+    let payment_amount = total_price
+        .checked_sub(lp_fee)
+        .ok_or(MMMErrorCode::NumericOverflow)?
+        .checked_sub(taker_fee as u64)
+        .ok_or(MMMErrorCode::NumericOverflow)?;
+
+    if payment_amount < args.min_payment_amount {
+        return Err(MMMErrorCode::InvalidRequestedPrice.into());
+    }
+
+    invoke_signed(
+        &system_instruction::transfer(buyside_sol_escrow_account.key, payer.key, payment_amount),
+        &[
+            buyside_sol_escrow_account.to_account_info(),
+            payer.to_account_info(),
+            system_program.to_account_info(),
+        ],
+        buyside_sol_escrow_account_seeds,
+    )?;
+
+    if lp_fee > 0 {
+        invoke_signed(
+            &system_instruction::transfer(buyside_sol_escrow_account.key, owner.key, lp_fee),
+            &[
+                buyside_sol_escrow_account.to_account_info(),
+                owner.to_account_info(),
+                system_program.to_account_info(),
+            ],
+            buyside_sol_escrow_account_seeds,
+        )?;
+    }
+    if referral_fee > 0 {
+        invoke_signed(
+            &system_instruction::transfer(
+                buyside_sol_escrow_account.key,
+                referral.key,
+                referral_fee,
+            ),
+            &[
+                buyside_sol_escrow_account.to_account_info(),
+                referral.to_account_info(),
+                system_program.to_account_info(),
+            ],
+            buyside_sol_escrow_account_seeds,
+        )?;
+    }
+
+    pool.lp_fee_earned = pool
+        .lp_fee_earned
+        .checked_add(lp_fee)
+        .ok_or(MMMErrorCode::NumericOverflow)?;
+    pool.spot_price = next_price;
+
+    try_close_escrow(
+        &buyside_sol_escrow_account.to_account_info(),
+        pool,
+        system_program,
+        buyside_sol_escrow_account_seeds,
+    )?;
+    try_close_sell_state(sell_state, payer.to_account_info())?;
+
+    // return the remaining per pool escrow balance to the shared escrow account
+    if pool.using_shared_escrow() {
```

Comment:
since we will need to consider shared escrow, which means we will need to use check_remaining_accounts_for_m2 and withdraw_m2 to withdraw the sol from M2 first

Reply:
oh good call!